src/core/batch.rkt

```diff
@@ -6,203 +6,104 @@
+    [_ expr]))
 
-  ; Translates programs into an instruction sequence of operations
-  (define (munge-ignore-approx prog)
-    (set! size (+ 1 size))
-    (match prog ; approx nodes are ignored
-      [(approx _ impl) (munge-ignore-approx impl)]
-      [_
-       (define node ; This compiles to the register machine
-         (match prog
-           [(list op args ...) (cons op (map munge-ignore-approx args))]
-           [_ prog]))
-       (hash-ref! exprhash
-                  node
-                  (lambda ()
-                    (begin0 (+ exprc varc) ; store in cache, update exprs, exprc
-                      (set! exprc (+ 1 exprc))
-                      (set! icache (cons node icache)))))]))
+;; Batches store these recursive structures, flattened
 
-  ; Translates programs into an instruction sequence of operations
-  (define (munge-include-approx prog)
-    (set! size (+ 1 size))
-    (define node ; This compiles to the register machine
-      (match prog
-        [(approx spec impl) (approx spec (munge-include-approx impl))]
-        [(list op args ...) (cons op (map munge-include-approx args))]
-        [_ prog]))
-    (hash-ref! exprhash
-               node
-               (lambda ()
-                 (begin0 (+ exprc varc) ; store in cache, update exprs, exprc
-                   (set! exprc (+ 1 exprc))
-                   (set! icache (cons node icache))))))
+(struct batch ([nodes #:mutable] [roots #:mutable] vars))
 
-  (define roots
-    (list->vector (map (if ignore-approx munge-ignore-approx munge-include-approx) exprs)))
-  (define nodes (list->vector (reverse icache)))
-  (define nodes-length (vector-length nodes))
+(define (batch-length b)
+  (cond
+    [(batch? b) (vector-length (batch-nodes b))]
+    [(mutable-batch? b) (hash-count (mutable-batch-index b))]
+    [else (error 'batch-length "Invalid batch" b)]))
 
-  (when timeline-push
-    (timeline-push! 'compiler (+ varc size) (+ exprc varc)))
-  (batch nodes roots vars nodes-length))
+(struct mutable-batch ([nodes #:mutable] [index #:mutable] [vars #:mutable]))
 
-(define (batch->progs batch)
-  (define roots (batch-roots batch))
-  (define nodes (batch-nodes batch))
+(define (make-mutable-batch)
+  (mutable-batch '() (make-hash) '()))
 
-  (define (unmunge reg)
-    (define node (vector-ref nodes reg))
-    (match node
-      [(approx spec impl) (approx spec (unmunge impl))]
-      [(list op regs ...) (cons op (map unmunge regs))]
-      [_ node]))
+(define (batch-push! b term)
+  (define hashcons (mutable-batch-index b))
+  (hash-ref! hashcons
+             term
+             (lambda ()
+               (let ([new-idx (hash-count hashcons)])
+                 (hash-set! hashcons term new-idx)
+                 (set-mutable-batch-nodes! b (cons term (mutable-batch-nodes b)))
+                 (when (symbol? term)
+                   (set-mutable-batch-vars! b (cons term (mutable-batch-vars b))))
+                 new-idx))))
```

Comment:
I like this function. I used to carry that weird `append-node` function around. This approach to insert instructions is better